src/mds/Server.cc

```diff
@@ -5821,13 +5823,191 @@ void Server::handle_remove_vxattr(MDRequestRef& mdr, CInode *cur)
   respond_to_request(mdr, -ENODATA);
 }
 
+const Server::XattrHandler Server::xattr_handlers[] = {
+  {
+    xattr_name: Server::DEFAULT_HANDLER,
+    description: "default xattr handler",
+    validate:  &Server::default_xattr_validate,
+    setxattr: &Server::default_setxattr_handler,
+    removexattr: &Server::default_removexattr_handler,
+  },
+  {
+    xattr_name: "ceph.mirror.info",
+    description: "mirror info xattr handler",
+    validate: &Server::mirror_info_xattr_validate,
+    setxattr: &Server::mirror_info_setxattr_handler,
+    removexattr: &Server::mirror_info_removexattr_handler
+  },
+};
+
+const Server::XattrHandler* Server::get_xattr_or_default_handler(std::string_view xattr_name) {
+  const XattrHandler *default_xattr_handler = nullptr;
+
+  for (auto &handler : xattr_handlers) {
+    if (handler.xattr_name == Server::DEFAULT_HANDLER) {
+      ceph_assert(default_xattr_handler == nullptr);
+      default_xattr_handler = &handler;
+    }
+    if (handler.xattr_name == xattr_name) {
+      dout(20) << "handler=" << handler.description << dendl;
+      return &handler;
+    }
+  }
+
+  ceph_assert(default_xattr_handler != nullptr);
+  dout(20) << "handler=" << default_xattr_handler->description << dendl;
+  return default_xattr_handler;
+}
+
+int Server::xattr_validate(CInode *cur, InodeStoreBase::xattr_map_const_ptr xattrs,
+                           const std::string &xattr_name, int op, int flags) {
+  if (op == CEPH_MDS_OP_SETXATTR) {
+    if (xattrs) {
+      if ((flags & CEPH_XATTR_CREATE) && xattrs->count(mempool::mds_co::string(xattr_name))) {
```

Comment:
```suggestion
      if ((flags & CEPH_XATTR_CREATE) && xattrs && xattrs->count(mempool::mds_co::string(xattr_name))) {
```

I think? Also for `RMXATTR` below. See these failures:

````
2020-10-07T05:41:15.102 INFO:tasks.ceph.mds.a.smithi198.stderr:*** Caught signal (Segmentation fault) **
2020-10-07T05:41:15.102 INFO:tasks.ceph.mds.a.smithi198.stderr: in thread 7f9c35f74700 thread_name:ms_dispatch
2020-10-07T05:41:15.102 INFO:tasks.ceph.mds.a.smithi198.stderr: ceph version 16.0.0-6209-gf0e17b88 (f0e17b8812dfdab73564df55d47fd1dafa530c61) pacific (dev)
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 1: /lib64/libpthread.so.0(+0x12dc0) [0x7f9c3d6f6dc0]
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 2: (std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > const, ceph::buffer::v15_2_0::ptr>, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > const, ceph::buffer::v15_2_0::ptr> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > >, mempool::pool_allocator<(mempool::pool_index_t)26, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > const, ceph::buffer::v15_2_0::ptr> > >::find(std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > const&) const+0x16) [0x55cbff60e346]
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 3: (Server::xattr_validate(CInode*, std::shared_ptr<std::map<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> >, ceph::buffer::v15_2_0::ptr, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > >, mempool::pool_allocator<(mempool::pool_index_t)26, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > const, ceph::buffer::v15_2_0::ptr> > > const>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)+0x1c0) [0x55cbff5ca3d0]
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 4: (Server::default_xattr_validate(CInode*, std::shared_ptr<std::map<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> >, ceph::buffer::v15_2_0::ptr, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > >, mempool::pool_allocator<(mempool::pool_index_t)26, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, mempool::pool_allocator<(mempool::pool_index_t)26, char> > const, ceph::buffer::v15_2_0::ptr> > > const>, Server::XattrOp*)+0x51) [0x55cbff5caa51]
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 5: (Server::handle_client_removexattr(boost::intrusive_ptr<MDRequestImpl>&)+0x2eb) [0x55cbff5cc76b]
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 6: (Server::dispatch_client_request(boost::intrusive_ptr<MDRequestImpl>&)+0xcdb) [0x55cbff5efa7b]
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 7: (MDCache::dispatch_request(boost::intrusive_ptr<MDRequestImpl>&)+0x33) [0x55cbff695003]
2020-10-07T05:41:15.103 INFO:tasks.ceph.mds.a.smithi198.stderr: 8: (MDSContext::complete(int)+0x56) [0x55cbff858386]
2020-10-07T05:41:15.104 INFO:tasks.ceph.mds.a.smithi198.stderr: 9: (void finish_contexts<std::vector<MDSContext*, std::allocator<MDSContext*> > >(ceph::common::CephContext*, std::vector<MDSContext*, std::allocator<MDSContext*> >&, int)+0x8d) [0x55cbff557ead]
2020-10-07T05:41:15.104 INFO:tasks.ceph.mds.a.smithi198.stderr: 10: (Locker::eval(CInode*, int, bool)+0x4d8) [0x55cbff7338e8]
2020-10-07T05:41:15.104 INFO:tasks.ceph.mds.a.smithi198.stderr: 11: (Locker::handle_client_caps(boost::intrusive_ptr<MClientCaps const> const&)+0x2048) [0x55cbff743a38]
2020-10-07T05:41:15.104 INFO:tasks.ceph.mds.a.smithi198.stderr: 12: (Locker::dispatch(boost::intrusive_ptr<Message const> const&)+0x104) [0x55cbff745634]
2020-10-07T05:41:15.104 INFO:tasks.ceph.mds.a.smithi198.stderr: 13: (MDSRank::handle_message(boost::intrusive_ptr<Message const> const&)+0xa7c) [0x55cbff55e95c]
2020-10-07T05:41:15.104 INFO:tasks.ceph.mds.a.smithi198.stderr: 14: (MDSRank::_dispatch(boost::intrusive_ptr<Message const> const&, bool)+0x7bb) [0x55cbff560eeb]
2020-10-07T05:41:15.104 INFO:tasks.ceph.mds.a.smithi198.stderr: 15: (MDSRankDispatcher::ms_dispatch(boost::intrusive_ptr<Message const> const&)+0x55) [0x55cbff5614e5]
2020-10-07T05:41:15.105 INFO:tasks.ceph.mds.a.smithi198.stderr: 16: (MDSDaemon::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x128) [0x55cbff551758]
2020-10-07T05:41:15.105 INFO:tasks.ceph.mds.a.smithi198.stderr: 17: (DispatchQueue::entry()+0x126a) [0x7f9c3ed3148a]
2020-10-07T05:41:15.105 INFO:tasks.ceph.mds.a.smithi198.stderr: 18: (DispatchQueue::DispatchThread::entry()+0x11) [0x7f9c3ede27c1]
2020-10-07T05:41:15.105 INFO:tasks.ceph.mds.a.smithi198.stderr: 19: /lib64/libpthread.so.0(+0x82de) [0x7f9c3d6ec2de]
2020-10-07T05:41:15.105 INFO:tasks.ceph.mds.a.smithi198.stderr: 20: clone()
````

/ceph/teuthology-archive/pdonnell-2020-10-07_03:27:03-fs-wip-pdonnell-testing-20201007.015751-distro-basic-smithi/5503131/teuthology.log

and these other jobs

`5 jobs: ['5503131', '5502947', '5503039', '5503084', '5502993']`

Reply:
That code block is already under `if (xattr) {... }`. However, the removexattr part does not have this check (the check existed before this refactor).

Reply:
Oh, you're right. It's likely the RMXATTR, yes.